cpmpy/transformations/to_bool.py

```diff
@@ -0,0 +1,317 @@
+
+            # The following encoding with the exactlyone constraint is then 
+            # equivalent to specifying x1 as an integer variable.
+            x1 = 4 * bv4 + 5 * bv5 + 6 * bv6 + 7 * bv7 + 8 * bv8
+    '''
+    constraints = []
+    ivarmap = {}
+    # Bool
+    if isinstance(int_var, _BoolVarImpl):
+        ivarmap[int_var] = int_var
+
+    # takes care of empty list!
+    elif isinstance(int_var, list):
+        for ivar in int_var:
+            sub_iv_mapping, int_cons = intvar_to_boolvar(ivar)
+            constraints += int_cons
+            ivarmap.update(sub_iv_mapping)
+
+    elif isinstance(int_var, NDVarArray):
+        lb, ub = int_var.flat[0].lb ,int_var.flat[0].ub
+        # reusing numpy notation if possible
+        bvs = boolvar(shape=int_var.shape + (ub - lb + 1,))
+
+        for i, ivar in np.ndenumerate(int_var):
+            ivarmap[ivar] = {ivar_val: bv for bv, ivar_val in zip(bvs[i], range(lb, ub+1))}
+            constraints.append(sum(bvs[i]) == 1)
+    else:
+        lb, ub = int_var.lb ,int_var.ub
+        bvs = boolvar(shape=(ub - lb +1))
+        ivarmap[int_var] = {ivar_val: bv for bv, ivar_val in zip(bvs, range(lb, ub+1))}
+
+        constraints.append(sum(bvs) == 1)
+
+    return ivarmap, constraints
+
+
+def to_bool_constraint(constraint, ivarmap):
+    '''
+        Decomposition of integer constraint using the provided mapping
+        of integer variables to boolvariables
+
+    :param constraint: Input flattened expression encoded in given boolean variable
+    encoding.
+    :type Expression:
+
+    :param ivarmap: Encoding of intvar to boolvar
+    :type dict
+
+    :return: bool: the computed output:
+        - True      if a solution is found (not necessarily optimal, e.g. could be after timeout)
+        - False     if no solution is found
+    '''
+    assert all(
+        True if iv in ivarmap else False for iv in get_variables(constraint)
+    ),f"""int var(s):
+        {[iv for iv in get_variables(constraint) if iv not in ivarmap]}
+    has not been mapped to a boolvar."""
+
+    bool_constraints = []
+
+    # CASE 1: True/False
+    if isinstance(constraint, bool):
+        return constraint
+
+    # CASE 2: Only bool vars in constraint
+    elif all(True if isinstance(arg, _BoolVarImpl) else False for arg in constraint.args):
+        return constraint
+
+    # CASE 3: Decompose list of constraints and handle individually
+    elif isinstance(constraint, (list, NDVarArray)):
+        for con in constraint:
+            bool_constraints += to_bool_constraint(con, ivarmap)
+
+    # CASE 4: base comparison constraints + ensure only handling what it can
+    elif isinstance(constraint, Comparison) and all(is_int(arg) or isinstance(arg, (_IntVarImpl, _BoolVarImpl)) for arg in constraint.args) :
+        bool_constraints += to_unit_comparison(constraint, ivarmap)
+
+    # CASE 5: global constraints
+    elif isinstance(constraint, (AllDifferent, AllEqual, Circuit, Table)):
```

Comment:
The way I see it is that, we want to replace in every constraint the intvar with its linear expression over Booleans.

Only, we also want to catch the special cases where it is a direct Comparison operator, and in that case replace the comparison by a direct expression on the Bools?

So, in any case, we should only 'do' something with constraints that contain an intvar, all others just pass through...
and the default should be to replace the intvar by a linexpr/wsum.
So then the question is, is this before or after flattening... I would say before!